Allow `(require ... :readers *)`

scauligi opened the pull request:
Closes #2265.

Breaks latest hyrule since the API for `hy.macros.require_reader` is slightly different, but it's a very easy fix.
I can submit a PR once this one goes through.

Kodiologist commented:
Are you sure it makes sense to unmangle in the error message? The mangled form is what shows up in `import` errors, for example.

The test uses `#[f[ … ]f]`, but has no replacement fields. Did you mean `#[=[ … ]=]` or so?

I'd recommend using `setv` instead of `let` when you can, at least in Hy itself, because history shows that `let`'s semantics are liable to change.

scauligi commented:
> unmangle in the error message?

To me it's more ergonomic to display the unmangled names in errors (and I would propose that we unmangle the names in the other import/require errors as well).

> f-string

Yup, simple copy-paste error. Fixed.

> `setv` instead of `let`

I hope `let`'s common-case semantics will be fairly stable from here out, but you do have a point. Changed.

Kodiologist commented:
> I would propose that we unmangle the names in the other import/require errors as well

I don't think we can, specifically for `import`, because such errors are raised at run time. Likewise for any other Python-level errors that involve variable names. So I think it's easier to understand if we always use mangled names.

scauligi commented:
Ahh, Python. Curses.

Kodiologist commented:
Sometimes it really be like that.

Kodiologist commented:
After that, we need a line in the documentation of `require` about the syntax `:readers *`. Then I think this will be good to go.

scauligi commented:
Done!

Kodiologist commented:
I still see `unmangle(name)` in `macros.py` line 181. Did you forget to push?

scauligi commented:
woops, I reverted the one for `:readers` but forgot that I had also changed normal `require`. Fixed now.

Kodiologist commented:
Thanks.
